RDKShell/RDKShell.cpp

```diff
@@ -7162,6 +7243,281 @@ namespace WPEFramework {
 
             return result;
         }
+
+        bool RDKShell::MemCheckpointRestoreClient::checkpoint(const std::string &callSign, uint32_t timeouteMs)
+        {
+            bool isCheckpointAllowed = false;
+
+            mProcessedAppsLock.lock();
+            if(mProcessedApps.find(callSign) == mProcessedApps.end())
+            {
+                //App not processed
+                isCheckpointAllowed = true;
+            }
+            else if(mProcessedApps[callSign] == RDKShell::MemCheckpointRestoreClient::RESTORING)
+            {
+                //App being processeed, then shedule checkpoint only if restore is ongoing
+                isCheckpointAllowed = true;
+            }
+
+            if(isCheckpointAllowed)
+            {
+                mProcessedApps[callSign] = RDKShell::MemCheckpointRestoreClient::CHECKPOINTING;
+            }
+
+            mProcessedAppsLock.unlock();
+
+            if(isCheckpointAllowed)
+            {
+                LOGINFO("Launching Checkpoint Thread for %s", callSign.c_str());
+                launchRequestThread(RDKShell::MemCheckpointRestoreClient::MEMCR_CHECKPOINT, callSign, timeouteMs);
+                return true;
+            }
+
+            LOGWARN("Checkpoint not allowed for %s", callSign.c_str());
+            return false;
+        }
+
+        bool RDKShell::MemCheckpointRestoreClient::restore(const std::string &callSign, uint32_t timeouteMs)
+        {
+            bool isRestoreAllowed = false;
+
+            mProcessedAppsLock.lock();
+            if(mProcessedApps.find(callSign) != mProcessedApps.end()
+                && (mProcessedApps[callSign] == RDKShell::MemCheckpointRestoreClient::CHECKPOINTING
+                        || mProcessedApps[callSign] == RDKShell::MemCheckpointRestoreClient::CHECKPOINTED))
```

Comment:
where do we set `mProcessedApps[callSign] = RDKShell::MemCheckpointRestoreClient::CHECKPOINTED`
or `mProcessedApps[callSign] = RDKShell::MemCheckpointRestoreClient::RESTORED` ? I could not find anywhere in the code ?

Reply:
yest, it set at the end of request thread lambda: https://github.com/rdkcentral/rdkservices/pull/3402/files#diff-57606d52dd6bdec85c0397fe933c537b2288904cebdc9ce65e8d8671046b875aR7368-R7387 , the transition is as follow:
CHECKPOINTING->CHECKPOINTED
RESTORING->remove from mProcessesApps